File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None

        merged = lists[0]
        lists.pop(0)
        if not merged:
            return None

        while len(lists) > 0:
            to_merge = lists[0]
            merged_pointer = merged
            head = None
            prev = None
            while merged_pointer and to_merge:
                less = None
                if merged_pointer.val < to_merge.val:
                    less = merged_pointer
                    merged_pointer = merged_pointer.next
                else:
                    less = to_merge
                    to_merge = to_merge.next

                if not head:
                    merged = less
                    head = less

                if prev:
                    prev.next = less
                
                prev = less
            
            if merged_pointer:
                prev.next = merged_pointer

            if to_merge:
                prev.next = to_merge

            lists.pop(0)

        return merged


        

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
 
public class Solution {
    public ListNode MergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null)
            return list2;
        if (list2 == null)
            return list1;

        var cur1 = list1;
        var cur2 = list2;
        ListNode head = null;
        ListNode prev = null;

        while (cur1 != null && cur2 != null)
        {
            ListNode less;
            if (cur1.val < cur2.val)
            {
                less = cur1;
                cur1 = cur1.next;
            }
            else
            {
                less = cur2;
                cur2 = cur2.next;
            }

            if (head == null)
                head = less;

            if (prev != null)
                prev.next = less;
            
            prev = less;
            
        }

        if (cur1 != null)
            prev.next = cur1;
        
        if (cur2 != null)
            prev.next = cur2;

        return head;
    }
}